File: MATLAB/example_15/ABC_binormal.m
function S = ABC_binormal(x)
% Generate samples from bivariate normal distribution

% Store local variables in memory
persistent n_pairs sigma id_1 id_2

% Load the data and define local variables - only once
if isempty(n_pairs)
    % Number of x,y pairs
    n_pairs = 10;
    % Define the sigma of the bivariate distribution
    sigma = 0.01^2 * eye(2);
    % Now reorganize the x-vector so that we get bivariate mu values
    id_1 = 1 : 2 : 2 * n_pairs - 1; id_2 = 2 : 2 : 2 * n_pairs;
end

% Check whether delta has been specified
if size(x,2) == 2 * n_pairs
    % Set delta to zero
    x = [ x 0 ];
end

% Define mu
mu = [ x(id_1)' x(id_2)' ];

% Now loop over each bivariate normal distribution
for i = 1 : n_pairs
    % Draw 50 points from each bivariate normal distribution
    X(i,1:2) = mean ( mvnrnd ( mu(i,1:2) , sigma, 50 ) + ...
        normrnd( 0 , x(end) , 50 , 2 ) );
end

% Now return vector with model simulated values
S = X(:);

File: Python/example_15/ABC_binormal.py
import numpy as np

############################
### Case study 15
############################
def ABC_binormal(x):
    # Persistent variables (stored globally for reuse)
    global n_pairs, sigma, id_1, id_2

    # Load the data and define local variables - only once
    if 'n_pairs' not in globals():
        n_pairs = 10                            # Number of x,y pairs
        sigma = 0.01**2 * np.eye(2)             # Define the sigma of the bivariate distribution
        # Reorganize the x-vector to get bivariate mu values
        id_1 = np.arange(0, n_pairs)            # Indices for x-coordinates
        id_2 = np.arange(n_pairs, 2 * n_pairs)  # Indices for y-coordinates

    # Check whether delta has been specified
    if len(x) == 2 * n_pairs:
        # Set delta to zero if it's not specified
        x = np.hstack([x, 0])

    # Define mu (mean of the bivariate normal distribution)
    mu = np.column_stack([x[id_1], x[id_2]])
    # Loop over each bivariate normal distribution
    X = []
    for i in range(n_pairs):
        # Draw 50 points from each bivariate normal distribution
        mvn_samples = np.random.multivariate_normal(mu[i, :2], sigma, 50)
        # Add a perturbation with std equal to delta [= 0]
        normal_samples = np.random.normal(0, x[-1], (50, 2))
        X.append(np.mean(mvn_samples + normal_samples, axis=0))

    # Return vector with model simulated values
    S = np.array(X).T.flatten()    
    # --> mu[0,0:2] ends up as S[0] and S[n_pairs]
    # --> mu[1,0:2] ends up as S[1] and S[n_pairs+1], etc.
    # --> this matches x coordinates
    # S = np.concatenate(X).flatten()

    return S
